infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Orchestrator.java

```diff
@@ -0,0 +1,153 @@
+package org.corfudb.infrastructure.orchestrator;
+
+import io.netty.channel.ChannelHandlerContext;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.format.Types;
+import org.corfudb.infrastructure.IServerRouter;
+import org.corfudb.protocols.wireprotocol.CorfuMsgType;
+import org.corfudb.protocols.wireprotocol.CorfuPayloadMsg;
+import org.corfudb.protocols.wireprotocol.orchestrator.AddNodeResponse;
+import org.corfudb.protocols.wireprotocol.orchestrator.OrchestratorRequest;
+import org.corfudb.protocols.wireprotocol.orchestrator.OrchestratorResponse;
+import org.corfudb.protocols.wireprotocol.orchestrator.QueryRequest;
+import org.corfudb.protocols.wireprotocol.orchestrator.QueryResponse;
+import org.corfudb.protocols.wireprotocol.orchestrator.Request;
+import org.corfudb.protocols.wireprotocol.orchestrator.Response;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Layout;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.corfudb.format.Types.OrchestratorRequestType.QUERY;
+
+/**
+ * The orchestrator is a stateless service that runs on all management servers and its purpose
+ * is to execute workflows. A workflow defines multiple smaller actions that must run in order
+ * that is specified by Workflow.getActions() to achieve a bigger goal. For example, growing the
+ * cluster. Initiated through RPC, the orchestrator will create a workflow instance and attempt
+ * to execute all its actions.
+ *
+ * <p>
+ * Created by Maithem on 10/25/17.
+ */
+
+@Slf4j
+public class Orchestrator {
+
+    final Callable<CorfuRuntime> getRuntime;
+    final Set<UUID> activeWorkflows = ConcurrentHashMap.newKeySet();
+
+    public Orchestrator(@Nonnull Callable<CorfuRuntime> runtime) {
+        this.getRuntime = runtime;
+    }
+
+    public void handle(@Nonnull CorfuPayloadMsg<OrchestratorRequest> msg,
+                       @Nonnull ChannelHandlerContext ctx,
+                       @Nonnull IServerRouter r) {
+
+        OrchestratorRequest orchReq = msg.getPayload();
+
+        if (orchReq.getRequest().getType() == QUERY) {
+            handleQuery(msg, ctx, r);
+        } else {
+            dispatch(msg, ctx, r);
+        }
+    }
+
+    void handleQuery(CorfuPayloadMsg<OrchestratorRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
+        QueryRequest req = (QueryRequest) msg.getPayload().getRequest();
+
+        Response resp;
+        if (activeWorkflows.contains(req.getId())) {
+            resp = new QueryResponse(true);
```

Comment:
does the queryResponse not return any status of the workflow or it's actions ?

Reply:
The only statuses that we care about are Active/Not active. This is because I don't keep track of failed workflows and ones that completed successfully.